### PR TITLE
ENH: OpenVMS port

### PR DESCRIPTION
**Pull request by vorfol**

### CODE REVIEW

**Comment by eric-wieser** [comment]
Does the OpenVMS compiler claim to support C99?

Roughly speaking, there seem to be two things  that this PR is trying to fix.

* Numpy makes assumptions about `size_t` vs `intptr_t` which the C standard does not require to be true. Fixing these is probably a good idea (but a lot of work), if we can do so without breaking the ABI.
* The OpenVMS compiler is not standard-compliant:
  * longdoubles treat `nan` incorrectly (IEEE-754 maybe applies?).
  * It reserves identifiers that are not labelled as reserved by the C standard.
 I'm wary of carrying around the maintenance burden for both a niche and non-conforming compiler.

File: numpy/core/memmap.py [hunk 1]
```diff
@@ -253,6 +254,7 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
                 fid.seek(bytes - 1, 0)
                 fid.write(b'\0')
                 fid.flush()
+                fsync(fid.fileno())
```
Comment (eric-wieser) [inline numpy/core/memmap.py:257]: What's the motivation for this change? Is it an improvement on other platforms too?

File: numpy/core/src/multiarray/nditer_constr.c [hunk 2]
```diff
@@ -2065,7 +2065,8 @@ npyiter_compute_index_strides(NpyIter *iter, npy_uint32 flags)
     if (NIT_ITERSIZE(iter) == 1) {
         if (itflags & NPY_ITFLAG_HASINDEX) {
             axisdata = NIT_AXISDATA(iter);
-            NAD_PTRS(axisdata)[nop] = 0;
+            // reset the index, it is just beyond the pointers
+            *(npy_intp*)&(NAD_PTRS(axisdata)[nop]) = 0;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/nditer_constr.c:2069]: I think an `NAD_INDEX` macro might be good to encapsulate this cast.

File: numpy/random/tests/test_extending.py [hunk 3]
```diff
@@ -57,9 +58,9 @@ def test_cython(tmp_path):
     # gh-16162: make sure numpy's __init__.pxd was used for cython
     # not really part of this test, but it is a convenient place to check
     with open(build_dir / 'extending.c') as fid:
-        txt_to_find = 'NumPy API declarations from "numpy/__init__.pxd"'
+        txt_to_find = re.compile(r'NumPy API declarations from "numpy\/__init__(\.([^.]*?))?.pxd"')
         for i, line in enumerate(fid):
-            if txt_to_find in line:
+            if txt_to_find.search(line):
                 break
```
Comment (eric-wieser) [inline numpy/random/tests/test_extending.py:64]: This looks like a bugfix for the new cython change, right? Perhaps we should create a new PR with this.

File: numpy/core/memmap.py (hunk 1 above)
Comment (vorfol) [reply numpy/core/memmap.py:257]: Maybe other platforms do synchronization inside flush(). But it is not obligatory. Without fsync() the next call of mmap() might fail. On OpenVMS it fails.

Comment (mattip) [reply numpy/core/memmap.py:257]: See point 9.3 in the [openvms faq](http://www.hoffmanlabs.org/vmsfaq/vmsfaq_013.html). Although that seems like an anti-feature, I would expect [`fflush`](https://linux.die.net/man/3/fflush) to be enough.

**Comment by charris** [comment]
You seem to have pulled in a lot of irrelevant changes. You can try `git rebase --no-ff main` to fix that. In general, rebasing is a cleaner way to go. There are other ways to fix things up if you get into trouble.
